Add new validation checks

gmgunter opened the pull request:
Closes #12

Add new checks against the frame-to-burst JSON file and phase-linked interferogram to the `disp-s1 validate` command:

* Checks that the spatial reference and bounding box of a dataset within the product matches the expected values for that frame from the frame-to-burst JSON file.
* Compares the unwrapped phase data for congruence with an interferogram file.

scottstanie commented on src/disp_s1/validate.py:
```diff
@@ -487,3 +523,201 @@ def compare(golden: Filename, test: Filename, data_dset: str = DSET_DEFAULT) ->
+    logger.info(f"Frame-to-burst JSON file: {json_file}")
+
+    # Extract the EPSG code and bounding box for this frame from the JSON file.
+    json_epsg, json_bbox = get_frame_bbox(frame_id=frame_id, json_file=json_file)
```
This has just reminded me that there's the redundant `get_frame_bbox`/other functions in disp-s1 that are now all in `opera-utils`... we should probably switch those over at some point

scottstanie replied:
(solved in https://github.com/opera-adt/disp-s1/pull/89)

scottstanie reviewed (approved):
Thanks Geoff 👍 looks grat
The only thing about this that comes to mind is more of a high-level algorithm question (@mirzaees )-

If we do a network unwrapping approach, what should `--igram` be? The equivalent single-reference interferogram (even if we didn't use that in the network)?

For single-reference-derived products this should work great; it might still be equivalent if we've solved an unwrapping network correctly for the integer ambiguities rather than a phase value

scottstanie commented:
1. Remembering a conversation with SDS, it would be good to add an option to let it run through all problems rather than exit out at the first error. I'm thinking something like `--fail-fast` as a flag, where the default is run through everything. Then it would print out at the end 

```
  File "/u/aurora-r0/staniewi/repos/disp-s1/src/disp_s1/validate.py", line 761, in validate
    raise ValidationError(combined_msg)
disp_s1.validate.ValidationError: Validation failures:
unwrapped phase dataset '/unwrapped_phase' failed validation: too many nan values (0.9532800448888682 > 0.01)
unwrapped phase dataset failed validation: phase values were not congruent with reference dataset
```

2. it wasn't changed in this PR, but the `nan_threashold=0.01` is way too low because of the geocoding tilt. a normal unw has like 20-30% nans

maybe these should be in another PR though

gmgunter commented:
I like the `--fail-fast` idea, but IMO let's make it a separate PR.

In the case where `--fail-fast` is disabled, would the individual validation errors contain traceback info? Or just the error message?

scottstanie commented:
> I like the `--fail-fast` idea, but IMO let's make it a separate PR.
> 
> In the case where `--fail-fast` is disabled, would the individual validation errors contain traceback info? Or just the error message?

The first version just had the error messages join by newlines- do you think the further traceback is helpful?

gmgunter commented:
I've rebased this PR and fixed the merge conflicts.

@scottstanie it might be worth taking a close second look since some of the conflicts were a bit hairy. Thanks!

I was able to successfully test the `disp-s1 validate` command using one of @mirzaees's test cases from the CalVal delivery on aurora.

```console
$ cd /u/aurora-r0/smirzaee/scratch/calval-delivery/delivery_data_small/
$ disp-s1 validate output/forward/20170217_20170430.nc --golden golden_output/forward/20170217_20170430.nc --igram scratch/forward/interferograms/20170217_20170430.int.tif
[INFO|validate|L537] 2024-08-08T15:14:45-0700: Comparing HDF5 contents...
[INFO|validate|L188] 2024-08-08T15:14:45-0700: Checking connected component labels...
[INFO|validate|L219] 2024-08-08T15:14:46-0700: Test unwrapped area: 68483111/69037976 (99.196%)
[INFO|validate|L220] 2024-08-08T15:14:46-0700: Reference unwrapped area: 68483111/69037976 (99.196%)
[INFO|validate|L221] 2024-08-08T15:14:46-0700: Intersection/Reference: 68483111/68483111 (100.000%)
[INFO|validate|L222] 2024-08-08T15:14:46-0700: Intersection/Union: 68483111/68483111 (100.000%)
[INFO|validate|L279] 2024-08-08T15:15:01-0700: Checking displacement...
[INFO|validate|L326] 2024-08-08T15:15:05-0700: Test nan count: 0/3026998 (0.000%)
[INFO|validate|L327] 2024-08-08T15:15:05-0700: Reference nan count: 0/3026998 (0.000%)
[INFO|validate|L401] 2024-08-08T15:15:08-0700: Mean absolute re-wrapped phase error: 0.00000 rad
[INFO|validate|L402] 2024-08-08T15:15:08-0700: Max absolute re-wrapped phase error: 0.00000 rad
[INFO|validate|L405] 2024-08-08T15:15:08-0700: Non-congruent pixel count: 0/3026998 (0.000%)
[INFO|validate|L449] 2024-08-08T15:15:08-0700: /identification/product_version: b'0.2' vs. b'0.2'
[INFO|validate|L449] 2024-08-08T15:15:10-0700: /metadata/disp_s1_software_version: b'0.4.1.post1.dev0+g318eeef.d20240806' vs. b'0.4.1.post1.dev0+g318eeef.d20240806'
[INFO|validate|L449] 2024-08-08T15:15:10-0700: /metadata/dolphin_software_version: b'0.26.0' vs. b'0.26.0'
[INFO|validate|L541] 2024-08-08T15:15:19-0700: Checking geospatial metadata...
[INFO|validate|L547] 2024-08-08T15:15:19-0700: Files /u/aurora-r0/smirzaee/scratch/calval-delivery/delivery_data_small/golden_output/forward/20170217_20170430.nc and /u/aurora-r0/smirzaee/scratch/calval-delivery/delivery_data_small/output/forward/20170217_20170430.nc match.
[INFO|validate|L571] 2024-08-08T15:15:19-0700: Checking for congruence with wrapped phase...
[INFO|validate|L587] 2024-08-08T15:15:21-0700: Interferogram file: /u/aurora-r0/smirzaee/scratch/calval-delivery/delivery_data_small/scratch/forward/interferograms/20170217_20170430.int.tif
[INFO|validate|L401] 2024-08-08T15:15:22-0700: Mean absolute re-wrapped phase error: 0.00014 rad
[INFO|validate|L402] 2024-08-08T15:15:22-0700: Max absolute re-wrapped phase error: 0.00087 rad
[INFO|validate|L405] 2024-08-08T15:15:22-0700: Non-congruent pixel count: 0/3026998 (0.000%)
[INFO|validate|L653] 2024-08-08T15:15:22-0700: Checking frame bounds again JSON file...
Downloading file 'opera-s1-disp-0.5.0-frame-to-burst.json.zip' from 'https://github.com/opera-adt/burst_db/releases/download/v0.5.0/opera-s1-disp-0.5.0-frame-to-burst.json.zip' to '/home/gunter/.cache/opera_utils/0.5.0'.
[INFO|validate|L662] 2024-08-08T15:15:23-0700: Dataset name: NETCDF:"/u/aurora-r0/smirzaee/scratch/calval-delivery/delivery_data_small/output/forward/20170217_20170430.nc":"//displacement"
[INFO|validate|L663] 2024-08-08T15:15:23-0700: Frame ID: 7091
[INFO|validate|L664] 2024-08-08T15:15:23-0700: Frame-to-burst JSON file: /home/gunter/.cache/opera_utils/0.5.0/opera-s1-disp-0.5.0-frame-to-burst.json.zip
[INFO|validate|L671] 2024-08-08T15:15:23-0700: Product EPSG code: 32612
[INFO|validate|L672] 2024-08-08T15:15:23-0700: Expected frame EPSG code: 32612
[INFO|validate|L682] 2024-08-08T15:15:23-0700: Product bounds: Bbox(left=280230.0, bottom=3555240.0, right=572310.0, top=3767970.0)
[INFO|validate|L683] 2024-08-08T15:15:23-0700: Expected frame bounds: Bbox(left=280230.0, bottom=3555240.0, right=572310.0, top=3767970.0)
```

scottstanie commented on src/disp_s1/validate.py:
```diff
@@ -358,17 +400,16 @@ def rewrap(phi: np.ndarray) -> np.ndarray:
 
-    atol_radians = atol * 4 * np.pi / wavelength
-    noncongruent_count = np.sum(abs_wrapped_diff > atol_radians)
+    noncongruent_count = np.sum(abs_wrapped_diff > atol)
```
We should just double check that I didn't compress too much from mantissa rounding so that we need a slightly lower tolerance. I think it should be fine though
